routemaster/server/endpoints.py

```diff
@@ -157,11 +157,21 @@ def create_label(state_machine_name, label_name):
     try:
         initial_state_name = \
             app.config.state_machines[state_machine_name].states[0].name
-        metadata = state_machine.create_label(app, label, initial_metadata)
+        if 'restore_label_and_restart' in data:
```

Comment:
Have we considered moving this to a separate endpoint? Creating a label is a considerably different operation to reinstating a deleted one and has rather different failure modes.

Reply:
I did consider it, but in my mind creating and restoring aren't really that different. Either way you're saying "I want this label to be a thing that I can track the state of", it's just in one case you have to also say "and I know it was previously a thing that I deleted, it's fine"

Happy to make it a `POST /{label}/restore` or `PUT /{label}` with `deleted=false` or something, if people disagree, though. I don't have hugely strong feelings.

Reply:
I think I agree with Peter that making this a separate API would be better...

- A failure means quite a different thing
- It's a destructive operation as we delete the original metadata (where both create and delete aren't really destructive)
- We still need the `restore_label_and_restart`, so we're still needing to track that the label previously existed in the client.

On the last point, I think that forcing the client to track this is probably desirable – if two run-throughs are _entirely_ independent at the application level then the label IDs should probably be different, whereas in this case styleme needs to be kinda careful about this, so making restoration more explicit will probably be handy.

`POST /{label}/restore` sounds good to me.

Reply:
> POST /{label}/restore sounds good to me.

Cool beans - I'm split it out into that 🙂 

> It's a destructive operation as we delete the original metadata (where both create and delete aren't really destructive)

Just to be clear, the delete operation is destructive. It deletes the metadata. The restore operation is overwriting what is guaranteed to be an empty dict.

Reply:
> Just to be clear, the delete operation is destructive. It deletes the metadata.

Oh! I had forgotten about that!

I think it's still good that we do a full replacement on restore anyway for safety, and I think it's still worth splitting out the API, but that's good to know. Sorry for the confusion!

Reply:
d559606ffd22efd5ba8f0a45b7257ffb7f48a6b5 🙂